### PR TITLE
[SourceEditor] Implemented other apporach to fix the update issue.

### PR DESCRIPTION
**Pull request by mkrueger**
This way it's a bit less complex.

### CODE REVIEW

**Comment by Therzok** [comment]
?w=1 looks better. LGTM.

**Comment by slluis** [comment]
Which issue is this fixing?

**Comment by mkrueger** [comment]
Highlighting update - the old fix broke webtools highlighting implementation.

In any case it's just a work around - we need a roslyn callback.

**Comment by mkrueger** [comment]
Basically reverts https://github.com/mono/monodevelop/commit/099391c3df9fc4e4fbe5c3aa6f5320336d6572bd

And implements another approach. But as said the parser should call the ui. Because of an update bug that used to work in older versions where we switched to the roslyn classifiers - there the editor often updated too much. With the drawing update optimization that has become visible. 

I need some feedback from the roslyn team how they handle that  the classifier API has no form of callbacks. The workaround is not the best solution for that - the parser knows about state changes.

**Comment by mkrueger** [comment]
It's already in the 15.6 PR:

https://github.com/mono/monodevelop/pull/4184

However having some days to try it out would be nice

File: main/src/addins/MonoDevelop.SourceEditor2/Mono.TextEditor/Gui/TextViewMargin.cs [hunk 1]
```diff
@@ -1356,39 +1356,32 @@ public ChunkDescriptor (TextDocument doc, DocumentLine line, int offset, int len
...
-					if (!LinesAreEqual(highlightedCurLine, updatedCurLine)) {
-						cachedLines = cachedLines.SetItem (curLineNumber, updatedCurLine);
-						await Runtime.RunInMainThread (delegate {
-							Document.CommitMultipleLineUpdate (lineNumber, lineNumber + (int)(textEditor.Allocation.Height / textEditor.LineHeight) + 1);
-						});
-						return;
-					}
+			var task = doc.SyntaxMode.GetHighlightedLineAsync (line, token);
+			if (task.IsCompleted) {
+				if (result != null && ShouldUpdateSpan (result, task.Result))
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.SourceEditor2/Mono.TextEditor/Gui/TextViewMargin.cs]:
```
       /// <summary>
        /// Gets whether this <see cref="Task">Task</see> has completed.
        /// </summary>
        /// <remarks>
        /// <see cref="IsCompleted"/> will return true when the Task is in one of the three
        /// final states: <see cref="System.Threading.Tasks.TaskStatus.RanToCompletion">RanToCompletion</see>,
        /// <see cref="System.Threading.Tasks.TaskStatus.Faulted">Faulted</see>, or
        /// <see cref="System.Threading.Tasks.TaskStatus.Canceled">Canceled</see>.
        /// </remarks>
```

This means we should check whether the task is faulted or cancelled before running this.

**Pull request merged**